pushy/src/test/java/com/turo/pushy/apns/proxy/HttpProxyHandlerFactoryTest.java

```diff
@@ -0,0 +1,150 @@
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import com.turo.pushy.apns.ApnsClientBuilder;
+
+public class HttpProxyHandlerFactoryTest {
+
+    private static class FixedProxiesSelector extends ProxySelector {
+
+        private final List<Proxy> availableProxies;
+
+        public FixedProxiesSelector(final Proxy... availableProxies) {
+            this.availableProxies = Arrays.asList(availableProxies);
+        }
+
+        @Override
+        public List<Proxy> select(URI uri) {
+            return availableProxies;
+        }
+
+        @Override
+        public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
+            // Not needed for this test.
+        }
+    }
+
+    private static class SingleHostHttpProxySelector extends ProxySelector {
+
+        private final String proxiedHost;
+
+        public SingleHostHttpProxySelector(final String proxiedHost) {
+            this.proxiedHost = proxiedHost;
+        }
+
+        @Override
+        public List<Proxy> select(URI uri) {
+            Proxy proxy;
+
+            if (uri.getHost().equals(proxiedHost)) {
+                proxy = DUMMY_HTTP_PROXY;
+            } else {
+                proxy = Proxy.NO_PROXY;
+            }
+
+            return Arrays.asList(proxy);
+        }
+
+        @Override
+        public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
+            // Not needed for this test.
+        }
+    }
+
+    private static final Proxy DUMMY_HTTP_PROXY = new Proxy(Proxy.Type.HTTP, new InetSocketAddress("httpproxy", 123));
+    private static final Proxy DUMMY_SOCKS_PROXY = new Proxy(Proxy.Type.SOCKS, new InetSocketAddress("socksproxy", 456));
+
+    private static ProxySelector defaultProxySelector;
+
+    @BeforeClass
+    public static void setUpClass() {
+        defaultProxySelector = ProxySelector.getDefault();
+    }
+
+    @AfterClass
+    public static void tearDownClass() {
+        ProxySelector.setDefault(defaultProxySelector);
+    }
+
+    @Test
+    public void testNoProxy()
+            throws URISyntaxException {
+
+        ProxySelector.setDefault(new FixedProxiesSelector(Proxy.NO_PROXY));
+
+        assertNull(HttpProxyHandlerFactory.fromSystemProxies());
+        assertNull(HttpProxyHandlerFactory.fromSystemProxies(ApnsClientBuilder.PRODUCTION_APNS_HOST));
+        assertNull(HttpProxyHandlerFactory.fromSystemProxies(ApnsClientBuilder.DEVELOPMENT_APNS_HOST));
+    }
+
+    @Test
+    public void testHasHttpProxy()
+            throws URISyntaxException {
+
+        ProxySelector.setDefault(new FixedProxiesSelector(DUMMY_HTTP_PROXY));
+
+        assertNotNull(HttpProxyHandlerFactory.fromSystemProxies());
+        assertNotNull(HttpProxyHandlerFactory.fromSystemProxies(ApnsClientBuilder.PRODUCTION_APNS_HOST));
+        assertNotNull(HttpProxyHandlerFactory.fromSystemProxies(ApnsClientBuilder.DEVELOPMENT_APNS_HOST));
+    }
+
+    @Test
+    public void testHasSocksProxy()
```

Comment:
Oh—and this test raises a question. If SOCKS proxies can he defined at a system level, should we think about:

1. Adding similar `fromSystemProxies` functionality to the SOCKS proxy handler factories?
2. Making a more general utility class that returns a `ProxyHandlerFactory`?

I recognize that both carry some baggage (in the latter case, how would a user indicate a preference if multiple types are available), but figure it's worth discussing while we're in here.

Reply:
I can't test SOCKS, so can only write such code blind. Happy to do so if you would like.

How you split up the static factory methods is a style thing. You could make a class to contain them. I prefer the current form because:

    HttpProxyHandlerFactory handler = HttpProxyHandlerFactory.fromSystemProxies(apnsHost)

seems pretty succinct. The alternative might be something like

    HttpProxyHandlerFactory handler = ProxyHandlerFactories.httpHandlerFromSystemProxies(apnsHost)

which feels a little more indirect to me. Happy to change it if you prefer though.